# Bank Fears Loanliness
This is my submission for the Coding Challenge at Hacker Earth. The problem is to figure out the probability that a member will default.

You can learn more about the challenge and the coding task here - https://www.hackerearth.com/challenge/competitive/machine-learning-challenge-one/machine-learning/bank-fears-loanliness/
Note: If the dexription id not available you can get the dataset from here (contains descriptions as well)-
https://www.dropbox.com/s/hhk58gf4zl0paco/Bank%20Fears%20Loanliness.zip?dl=0

In [1]:
# importing basic libraries we need to do our analysis
import pandas as pd
import numpy as np
# sklearn library for machine learning
from sklearn.ensemble import RandomForestRegressor
from sklearn.feature_extraction import DictVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

ImportError: DLL load failed: %1 is not a valid Win32 application.

In [2]:
# Let's start by reading in the data file and understanding what our data looks like.
train_df = pd.read_csv("train_indessa.csv")
test_df = pd.read_csv("test_indessa.csv")
train_df

member_id  loan_amnt  funded_amnt  funded_amnt_inv       term  \
0        58189336      14350        14350         14350.00  36 months   
1        70011223       4800         4800          4800.00  36 months   
2        70255675      10000        10000         10000.00  36 months   
3         1893936      15000        15000         15000.00  36 months   
4         7652106      16000        16000         16000.00  36 months   
5        10247268      15000        15000         14950.00  36 months   
6         8089625       5000         5000          4975.00  36 months   
7        23043116       6000         6000          6000.00  36 months   
8        45900933       6000         6000          6000.00  36 months   
9        41272507      34550        34550         34550.00  60 months   
10       56621560      25000        25000         24975.00  36 months   
11        4221941      15000        15000         15000.00  36 months   
12       43387539      18000        18000         18000.00  36 months   
13       23001903       9000         9000          9000.00  36 months   
14       13670372       8800         8800          8800.00  36 months   
15       49874851       8000         8000          8000.00  36 months   
16        9878144      18450        18450         18450.00  60 months   
17       12866989      30750        30750         30700.00  36 months   
18       58568833       9000         9000          9000.00  36 months   
19       62725253      24000        24000         24000.00  36 months   
20       66791957      20000        20000         20000.00  60 months   
21        1699807      12400        12400         12400.00  36 months   
22       54486146       4200         4200          4200.00  36 months   
23       41408977      10500        10500         10500.00  36 months   
24       63047232      16000        16000         16000.00  36 months   
25       22741274      23800        23800         23800.00  60 months   
26       17563299       7200         7200          7150.00  36 months   
27       10317534      19200        19200         19175.00  36 months   
28       50609137      15000        15000         15000.00  36 months   
29       70008968       4000         4000          4000.00  36 months   
...           ...        ...          ...              ...        ...   
532398   23543780      18000        18000         18000.00  36 months   
532399   16543128       4700         4700          4700.00  36 months   
532400   73414202      20000        20000         20000.00  36 months   
532401   54776171       7800         7800          7800.00  36 months   
532402   52377140      10400        10400         10400.00  36 months   
532403   59468945      22375        22375         22375.00  60 months   
532404   68127956       8000         8000          8000.00  36 months   
532405    5906972      18000        18000         18000.00  60 months   
532406   70600730      10000        10000         10000.00  36 months   
532407   59780170      16000        16000         16000.00  36 months   
532408   69555199       7200         7200          7200.00  36 months   
532409    4583654      17625        17625         17625.00  36 months   
532410     272515       9700         9700          8283.54  36 months   
532411   10340541      20000        20000         20000.00  36 months   
532412   48073747       7200         7200          7200.00  36 months   
532413   12956724      19200        19200         19200.00  60 months   
532414   26517526      20000        20000         20000.00  36 months   
532415     965370      10000        10000         10000.00  36 months   
532416   19786238       1800         1800          1800.00  36 months   
532417   68700956      10000        10000         10000.00  36 months   
532418   71054913      10000        10000         10000.00  36 months   
532419   51774468       4800         4800          4800.00  36 months   
532420   16661491       5700         5700          5700.00  36 mon

In [3]:
#As we can see the dataset is huge and there are a lot of rows/columns in it. 
#We look at the shape of the data and later try to understand what these columns mean.
train_df.shape

(532428, 45)

### Cleaning the dataset
Our training dataset has 532428 rows and 45 columns. Which means that we have 532427 instances or different records for the customers of our bank. The 45 columns means that we have 45 different features or measurements for each of our instances. 

The main step of any data science or machine learning problem is setting up the data in such a way that we it can be used in our machine learning model to yield the highest accuracy. In other words, feature extraction and selection and data tyding plays more important role than the machine learning model/algorithm we use.
Thus, we are looking at the unique values in each column to understand the sparcity of the data and to drop all the columns which are unnecessary

In [4]:
# basic exploration of a column of our dataset. Looking at three different things
def studyColumn(df, columnName):
    # different ways of looking at unique values in a column
    #df["member_id"].unique()
    print("Unique values in column are: ", df[columnName].unique())
    # counting the total number of null or nan values
    print("Total number of null values is: ", df[columnName].isnull().sum())
    # counting th  total number of each type of value in a column
    print("The frequency of each unique value is: ", df[columnName].value_counts(normalize = True))

In [5]:
studyColumn(train_df, "emp_length")

Unique values in column are:  ['9 years' '< 1 year' '2 years' '10+ years' '5 years' '8 years' '7 years'
 '4 years' 'n/a' '1 year' '3 years' '6 years']
Total number of null values is:  0
The frequency of each unique value is:  10+ years    0.328880
2 years      0.088793
< 1 year     0.079359
3 years      0.079213
1 year       0.064238
5 years      0.062718
4 years      0.059315
n/a          0.050506
7 years      0.050110
8 years      0.049665
6 years      0.048346
9 years      0.038856
Name: emp_length, dtype: float64


In [6]:
# Dropping columns

'''As we can see in our dataset we have a lot of empty columns and a lot of columns with 0 values or 
90% - 95% of 0s. There are various methods to fill the empty values, however for the first dry run and to just get a 
baseline for future reference we will simply ignore/drop these columns to create our first model'''

def droppingCols(df):
    # stripping " months" from each value in column 'term'
    df['term'] = df['term'].map(lambda x: x.rstrip(' months'))
    #converting to int
    df['term'] = pd.to_numeric(df['term'], errors='ignore')
    # multiplying by 4 to convert to weeks
    df['term'] *= 4
    # removed "th week" from each in value in 'last_week_pay', already in weeks so no need to convert
    df['last_week_pay'] = df['last_week_pay'].map(lambda x: x.rstrip('th week'))

    # Removing < and +years from all the emp_length values
    df['emp_length'] = df['emp_length'].map(lambda x: x.rstrip('+ years, year'))
    # Replacing n/ with np.null for further processing And changing < 1 = 0
    df['emp_length'].replace(to_replace = "< 1", value = 0, inplace = True)
    df['emp_length'].replace(to_replace = "n/", value = np.nan, inplace = True)

    # from our exploration we have selected the following columns to remove from our dataset
    columnToRemove = ['batch_enrolled', 'grade', 'sub_grade', 'emp_title', 'verification_status', 'desc', 
                      'purpose', 'title', 'zip_code', 'addr_state', 'mths_since_last_delinq', 'mths_since_last_record', 
                      'total_rec_late_fee', 'mths_since_last_major_derog', 'verification_status_joint']
    # only do once in each run as it will give an error if done without loading the dataset again (columns already removed)
    df.drop(columnToRemove, axis = 1, inplace = True)
    df.shape
    return df

In [7]:
# Creating a new column in our df
# money_bank_paid = funded_amnt - funded_amnt_inv (Substracting in this way to not have negative values)
def addCols(df):
    df['money_bank_paid'] = df['funded_amnt'] - df['funded_amnt_inv']

    # Converting string based abstract classes to category type
    df['home_ownership'] = df['home_ownership'].astype('category')
    df['pymnt_plan'] = df['pymnt_plan'].astype('category')
    df['initial_list_status'] = df['initial_list_status'].astype('category')
    df['application_type'] = df['application_type'].astype('category')

    # all columsn with dtype category.
    categoryColumns = df.select_dtypes(['category']).columns
    df[categoryColumns] = df[categoryColumns].apply(lambda x: x.cat.codes)

    # way to convert a single column
    #df['home_ownership'] = df['home_ownership'].cat.rename_categories(range(len(df['home_ownership'].cat.categories)))
    df.shape
    return df

In [8]:
def replaceEmpty(df):
    # replacing the empty spaces with nan values and nan with 0
    df = df.replace(r'\s+', np.nan, regex=True)
    df = df.replace(np.nan, 0, regex=True)
    df = df.replace('NA', 0, regex=True)
    return df

In [9]:
# calling the changes on both the dataframes
# first for the trainig dataset
train_df = droppingCols(train_df)
train_df = addCols(train_df)
train_df = replaceEmpty(train_df)

test_df = droppingCols(test_df)
test_df = addCols(test_df)
test_df = replaceEmpty(test_df)


### Building the Model
Now that we've our dataset cleaned and the empty and null values dealt with, we can go ahead and build our first model with this new dataset and get a baseline for accuracy and predictions. We will use Random forest for the first model we're going to build. Random Trees are a form of decision tress and you can read more about them here - 
https://en.wikipedia.org/wiki/Random_forest

In [10]:
# removing extra indexes from our Y_test and Y_train
def dropIndex(df):
    df = df.reset_index(level=None, drop=False, name=None, inplace=False)
    #df.drop('index', axis = 1, inplace = True)
    #df.set_index('loan_status', drop=True, append=False, inplace=True, verify_integrity=False)
    return df

In [11]:
# our training dataset split in to x and y
Y_train = train_df['loan_status']
X_train = train_df.drop('loan_status', axis = 1)

In [12]:
# We are using the random forest for this problem
RFmodel = RandomForestRegressor(n_estimators=10)
# fitting the model on the training dataset
RFmodel.fit(X_train, Y_train)

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_split=1e-07, min_samples_leaf=1,
           min_samples_split=2, min_weight_fraction_leaf=0.0,
           n_estimators=10, n_jobs=1, oob_score=False, random_state=None,
           verbose=0, warm_start=False)

In [13]:
# Now that we've our model we can test it on our test dataset for predicting the results
Y_pred = RFmodel.predict(test_df)

In [14]:
# Creating a temp df to write the results in to a csv for submission
temp = pd.DataFrame(test_df['member_id'])
temp['loan_status'] = Y_pred
temp.reset_index(level = None, drop = True, inplace = True)
temp['loan_status'].replace(0, 0.01,inplace=True)
temp['loan_status'].replace(1, 0.99,inplace=True)
temp.to_csv("submission.csv", sep = ",", index = False)
temp

member_id  loan_status
0        11937648         0.01
1        38983318         0.01
2        27999917         0.01
3        61514932         0.01
4        59622821         0.01
5        28822038         0.01
6        10718089         0.90
7        58114582         0.01
8        35023176         0.80
9         1268247         0.01
10       15559751         0.99
11       41234300         0.80
12       53958774         0.01
13        8074949         0.01
14       67601509         0.01
15       65560239         0.01
16       14767770         0.10
17        1726792         0.01
18        9899439         0.01
19       15120294         0.01
20       12019444         0.01
21       35242810         0.01
22        1057101         0.90
23       23024164         0.01
24       45430320         0.90
25       46396580         0.10
26       58832624         0.01
27       14990008         0.01
28       57738287         0.01
29       66382471         0.01
...           ...          ...
354921   19848222         0.01
354922   58710103         0.10
354923   71207833         0.01
354924   61595559         0.20
354925   59891936         0.01
354926   55781739         0.20
354927   49353323         0.01
354928     673495         0.70
354929   28653239         0.01
354930   45471139         0.01
354931    5329126         0.99
354932   22381672         0.01
354933   65005753         0.01
354934   35873497         0.01
354935   47994257         0.01
354936    3874064         0.20
354937   68780700         0.01
354938   10976807         0.01
354939   55096221         0.01
354940   13600061         0.70
354941   55437336         0.01
354942   25856020         0.01
354943   10386922         0.01
354944   70721678         0.01
354945   70348484         0.01
354946   19145105         0.01
354947   46304777         0.01
354948     903745         0.80
354949   53032475         0.01
354950     994245         0.80

[354951 rows x 2 columns]